milli/tests/search/query_criteria.rs

```diff
@@ -0,0 +1,531 @@
+
+    assert_eq!(documents_ids, expected_external_ids);
+}
+
+#[test]
+fn test_asc_on_unexisting_field_should_return_all_1() {
+    let criteria = vec![Criterion::Asc(S("unexisting_field"))];
+    let index = search::setup_search_index_with_criteria(&criteria);
+    let mut rtxn = index.read_txn().unwrap();
+
+    let mut search = Search::new(&mut rtxn, &index);
+    search.query(search::TEST_QUERY);
+    search.limit(EXTERNAL_DOCUMENTS_IDS.len());
+    search.optional_words(false);
+    search.authorize_typos(false);
+
+    let SearchResult { matching_words: _matching_words, candidates: _candidates, documents_ids } = search.execute().unwrap();
+
+    let criteria = vec![];
+    let expected_external_ids: Vec<_> = search::expected_order(&criteria, false, false).into_iter().map(|d| d.id).collect();
+    let documents_ids = search::internal_to_external_ids(&index, &documents_ids);
+
+    assert_eq!(documents_ids, expected_external_ids);
+}
+
+#[test]
+fn test_desc_on_unexisting_field_should_return_all_2() {
+    let criteria = vec![Criterion::Desc(S("unexisting_field"))];
+    let index = search::setup_search_index_with_criteria(&criteria);
+    let mut rtxn = index.read_txn().unwrap();
+
+    let mut search = Search::new(&mut rtxn, &index);
+    search.query(search::TEST_QUERY);
+    search.limit(EXTERNAL_DOCUMENTS_IDS.len());
+
+    let SearchResult { matching_words: _matching_words, candidates: _candidates, documents_ids } = search.execute().unwrap();
+
+    let criteria = vec![];
+    let expected_external_ids: Vec<_> = search::expected_order(&criteria, true, true).into_iter().map(|d| d.id).collect();
+    let documents_ids = search::internal_to_external_ids(&index, &documents_ids);
+
+    assert_eq!(documents_ids, expected_external_ids);
+}
+
+#[test]
+fn test_asc_on_unexisting_field_should_return_all_2() {
+    let criteria = vec![Criterion::Asc(S("unexisting_field"))];
+    let index = search::setup_search_index_with_criteria(&criteria);
+    let mut rtxn = index.read_txn().unwrap();
+
+    let mut search = Search::new(&mut rtxn, &index);
+    search.query(search::TEST_QUERY);
+    search.limit(EXTERNAL_DOCUMENTS_IDS.len());
+
+    let SearchResult { matching_words: _matching_words, candidates: _candidates, documents_ids } = search.execute().unwrap();
+
+    let criteria = vec![];
+    let expected_external_ids: Vec<_> = search::expected_order(&criteria, true, true).into_iter().map(|d| d.id).collect();
+    let documents_ids = search::internal_to_external_ids(&index, &documents_ids);
+
+    assert_eq!(documents_ids, expected_external_ids);
+}
+
+#[test]
+fn criteria_mixup() {
+    use Criterion::*;
+    let index = search::setup_search_index_with_criteria(&vec![Words, Attribute, Desc(S("asc_desc_rank")), Exactness, Proximity, Typo]);
+
+
+    let criteria_mix = {
+        let desc = || Desc(S("asc_desc_rank"));
+        // all possible criteria order
+        vec![
```

Comment:
Wouldn't it be possible to use the itertools [combinations](https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.combinations) or maybe [permutations](https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.permutations) methods instead?